Add the ability to upload llanfair runs

BatedUrGonnaDie opened the pull request.

glacials commented on app/models/run.rb:
```diff
@@ -35,7 +35,7 @@ class Run < ActiveRecord::Base
   class << self
     def programs
-      [Urn, LiveSplit, SplitterZ, TimeSplitTracker, WSplit]
+      [Urn, LiveSplit, Llanfair, SplitterZ, TimeSplitTracker, WSplit]
```
Put Llanfair in at the beginning, I like to sort this by implementation date

glacials commented on app/models/run_file.rb:
```diff
@@ -6,16 +6,24 @@ class RunFile < ActiveRecord::Base
   def self.for_file(file)
     if file.respond_to?(:read)
-      RunFile.for_text(file.read)
+      RunFile.for_text(file.read, File.binary?(file.tempfile))
```
This breaks tests for me:

```bash
⇒  rspec spec/models/run_spec.rb
# ...
/Users/glacials/pj/splits-io/app/models/run_file.rb:9:in `for_file': undefined method `tempfile' for #<File:0x007f97e9928190> (NoMethodError)
	from /Users/glacials/pj/splits-io/spec/factories/run_factory.rb:3:in `block (2 levels) in <top (required)>'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/factory_girl-4.5.0/lib/factory_girl/syntax/default.rb:18:in `instance_eval'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/factory_girl-4.5.0/lib/factory_girl/syntax/default.rb:18:in `factory'
	from /Users/glacials/pj/splits-io/spec/factories/run_factory.rb:2:in `block in <top (required)>'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/factory_girl-4.5.0/lib/factory_girl/syntax/default.rb:49:in `instance_eval'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/factory_girl-4.5.0/lib/factory_girl/syntax/default.rb:49:in `run'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/factory_girl-4.5.0/lib/factory_girl/syntax/default.rb:7:in `define'
	from /Users/glacials/pj/splits-io/spec/factories/run_factory.rb:1:in `<top (required)>'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/bundler/gems/rails-ee372bdcebf9/activesupport/lib/active_support/dependencies.rb:274:in `require'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/bundler/gems/rails-ee372bdcebf9/activesupport/lib/active_support/dependencies.rb:274:in `block in require'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/bundler/gems/rails-ee372bdcebf9/activesupport/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/bundler/gems/rails-ee372bdcebf9/activesupport/lib/active_support/dependencies.rb:274:in `require'
	from /Users/glacials/pj/splits-io/spec/rails_helper.rb:7:in `block in <top (required)>'
	from /Users/glacials/pj/splits-io/spec/rails_helper.rb:7:in `each'
	from /Users/glacials/pj/splits-io/spec/rails_helper.rb:7:in `<top (required)>'
	from /Users/glacials/pj/splits-io/spec/models/run_spec.rb:1:in `require'
	from /Users/glacials/pj/splits-io/spec/models/run_spec.rb:1:in `<top (required)>'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `load'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `block in load_spec_files'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `each'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `load_spec_files'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:96:in `setup'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:84:in `run'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:69:in `run'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:37:in `invoke'
	from /Users/glacials/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.1.7/exe/rspec:4:in `<top (required)>'
	from /Users/glacials/.rbenv/versions/2.2.2/bin/rspec:23:in `load'
	from /Users/glacials/.rbenv/versions/2.2.2/bin/rspec:23:in `<main>'
```

BatedUrGonnaDie replied:
tempfile is where the actual file gets uploaded to so that I can check if it is a binary file since File.binary?() needs a file path to check.  I'll look and see if there's a way to avoid doing this.

glacials commented on app/models/run_file.rb:
```diff
@@ -6,16 +6,24 @@ class RunFile < ActiveRecord::Base
+  def self.for_text(file_text, binary)
     digest = Digest::SHA256.hexdigest(file_text)
-    where(digest: digest).first_or_create(file: file_text)
+    where(digest: digest).first_or_create(file: binary ? RunFile.unpack_binary(file_text) : file_text)
```
I'd prefer the binary stuff be in a separate `for_binary` method so that we're not calling this one with a sometimes-unused argument. Then `for_file` can determine which method it needs to call.

glacials commented on app/controllers/runs_controller.rb:
```diff
@@ -76,7 +76,7 @@ def download
     send_data(
       if params[:program] == @run.program.to_s
-        @run.file
+        @run.program.to_s == "llanfair" ? RunFile.pack_binary(@run.file) : @run.file
```
I'd like to avoid adding filetype logic to the controller. Instead of deciding whether to pack here, it'd be nicer to put a method on `Run` like 

```ruby
def original_file
  program == :llanfair ? RunFile.pack_binary(file) : file
end
```

then just be able to call `@run.original_file` from here.

glacials commented on app/models/llanfair.rb:
```diff
@@ -0,0 +1,113 @@
+
+      file.seek(1, 1)
+      title_len = file.read(2).unpack("n")[0]
+      run[:title] = file.read(title_len)
```
Don't bother setting a `run[:title]`, we just call it `run[:name]` no matter what the program likes to call it, so that runs from different programs have mostly the same structure after parsing.

glacials commented on app/models/llanfair.rb:
```diff
@@ -0,0 +1,113 @@
+      file = StringIO.new(file)
+      file.seek(197, 1)
+      goal_len = file.read(2).unpack("n")[0]
+      run[:goal] = file.read(goal_len)
```
Hmm, it's just now hit me that Llanfair has both a Title and a Goal. What's the typical use of these fields for Llanfair users? I imagine Goal is mostly "`gamename` `categoryname`"? Or is Title typically `gamename` and Goal is typically `categoryname`?

The three standard fields for inputs we use are `run[:game]`, `run[:category]`, and `run[:name]` (with `run[:name]` defaulted to "`run[:game]` `run[:category]`" for some timers), so it's worth thinking about if the conventional uses for Title or Goal fit into two of those three slots.

BatedUrGonnaDie replied:
Title is typically game + category, and goal is usually a time, which is auto set to be your pb time if you do not change it.  Llanfair does allow you to set it to whatever you want, so you could set it to be something like "Beat X's time" if you wanted.  I wasn't completely sure if it was a good idea to include that in the name because it could just be a time.

Llanfair has some weird quirks to it that make setting a name hard since you have so few input areas to work with.

glacials replied:
:+1: well explained.

glacials commented:
Sweet, amazing. looking forward to getting this in.

BatedUrGonnaDie commented:
When I get home I'll implement the changes.

BatedUrGonnaDie commented:
Ok, made llanfair first in the list, fixed the test by using file.path() instead of using the tempfile instance variable, moved stuff to for_binary, moved the program checking to run model, and ditched run[:title] because I couldn't think of a good way to incorporate the goal in with it so for now I feel it's best to leave it out of the name.

glacials commented:
Sweet, I'm ready to merge after one more change: Can you remove the "Why no Llanfair?" question from `app/views/pages/faq.slim`? :smile:

BatedUrGonnaDie commented:
Good to go.
